### PR TITLE
[FR] Implement TryParse for FieldValue

### PR DESCRIPTION
**Pull request by maririos**
We have been receiving feedback from the community that they will like for our `FieldValue` class to have a `TryParse` method so their code is not handling so many exceptions (try/catch blocks), This PR attempts to do that and is using Sample Invoice to show how the behavior will be.

The main concern users had was that they thought checking the `type` of the `FieldValue` was enough for trying our `AsType` methods. When they got exceptions thrown, there was confusion on the why.
After looking into it, our `AsType` methods also verifies that the service successfully returned a value under `ValueType` property. In the cases the service doesn't return it, we should be throwing.

Sadly, this is not true for all cases:
- `AsString` returns null instead of throwing when `ValueString` is not returned.
- `AsFloat` has a bug where it tries to parse the text and if it doesn't work it throws. That will be fixed in another issue.

This draft is to validate how this approach is going before I get too deep into all implementation.

### CODE REVIEW

File: sdk/formrecognizer/Azure.AI.FormRecognizer/src/FieldValue.cs [hunk 1]
```diff
@@ -24,7 +24,36 @@ internal FieldValue(FieldValue_internal fieldValue, IReadOnlyList<ReadResult> re
         }
 
         /// <summary>
-        /// Initializes a new instance of the <see cref="FieldValue"/> structure.
+        /// Initializes a new instance of the <see cref="FieldValue"/> structure. Used for mocking purposes.
+        /// </summary>
+#pragma warning disable CA1801 // Review unused parameters
+        internal FieldValue(FieldValueType type, string valueString, DateTimeOffset? valueDate, TimeSpan? valueTime, string valuePhoneNumber, float? valueNumber, long? valueInteger, IReadOnlyList<FieldValue_internal> valueArray, IReadOnlyDictionary<string, FieldValue_internal> valueObject, SelectionMarkState? valueSelectionMark, FieldValueGender? valueGender, string valueCountry, string text)
```
Comment (maririos) [inline sdk/formrecognizer/Azure.AI.FormRecognizer/src/FieldValue.cs:30]: I need a method that gives the ability to pass type and text and not necessarily the valueType

File: sdk/formrecognizer/Azure.AI.FormRecognizer/src/FieldValue.cs [hunk 2]
```diff
@@ -442,5 +471,154 @@ public FieldValueGender AsGender()
...
+            {
+                case FieldValueType.String:
+                    try
+                    {
+                        result = (T)(object)AsString();
+                        if (result == null)
+                        {
+                            result = default;
+                            parse = false;
+                        }
```
Comment (maririos) [inline sdk/formrecognizer/Azure.AI.FormRecognizer/src/FieldValue.cs:493]: In the case of strings, `AsString` will return `null` if there is no value in `ValueString`. I wanted to change it to throw an exception, but this will be a breaking change from GA version... and because it is a string, not sure it is worth it.
But when talking about `TryParse` if the value is `null`, then in my head it means it didn't work, so I prefer to return `false` and `null`.
- Triple checking, fixing `AsString` to throw an exception if `ValueString` is not provided, while before we didn't and just returned null, is a breaking change, right?
- Do you think this is inconsistent? 
- Should I follow the behavior of as `AsString` and let `TryParse` return `true` and `null`?

File: sdk/formrecognizer/Azure.AI.FormRecognizer/src/FormRecognizerModelFactory.cs [hunk 3]
```diff
@@ -175,9 +175,55 @@ public static FieldData FieldData(FieldBoundingBox boundingBox, int pageNumber,
...
+        {
+            return new FieldValue(
+                type,
+                valueString,
+                valueDate,
+                valueTime,
+                valuePhoneNumber,
+                valueNumber,
+                valueInteger,
+                default,
```
Comment (maririos) [inline sdk/formrecognizer/Azure.AI.FormRecognizer/src/FormRecognizerModelFactory.cs:213]: At this point, I will need to implement a way to pass from `FormField` to `FieldValue_internal` so the user can use this mocking method.
But that is besides this effort, so will do later

File: sdk/formrecognizer/Azure.AI.FormRecognizer/tests/samples/Sample13_RecognizeInvoicesFromFile.cs [hunk 4]
```diff
@@ -42,102 +42,143 @@ public async Task RecognizeInvoicesFromFile()
...
             {
-                if (vendorNameField.Value.ValueType == FieldValueType.String)
+                if (vendorNameField.Value.TryParse(out string vendorName))
                 {
-                    string vendorName = vendorNameField.Value.AsString();
                     Console.WriteLine($"Vendor Name: '{vendorName}', with confidence {vendorNameField.Confidence}");
                 }
+                else
+                {
+                    Console.WriteLine($"Not able to parse Vendor Name. Consider using 'ValueData.Text' property which current value is '{vendorNameField.ValueData.Text}'");
```
Comment (maririos) [inline sdk/formrecognizer/Azure.AI.FormRecognizer/tests/samples/Sample13_RecognizeInvoicesFromFile.cs:51]: confirming with Service team that message is appropiate

**Comment by tg-msft** [comment]
I worry a little that `TryParse` implies things that this API isn't doing.  If I have a string field with the value `"10"`, I don't think `TryParse(out int x)` will actually parse the value and give me back an `int`.

What if we made it possible for `null` to be returned from variants of all your `AsType()` methods? I could imagine something like `long? number = fieldVal.AsInt64OrNull();` or even `long? number = fieldVal.AsPhoneNumber(/* throwOnError = */ false);`.

**Comment by maririos** [comment]
>I worry a little that TryParse implies things that this API isn't doing. If I have a string field with the value "10", I don't think TryParse(out int x) will actually parse the value and give me back an int.

Great point. Never thought of that

>long? number = fieldVal.AsInt64OrNull()

I really like this proposal. it will allow customers to write blocks where if the value is null, consider using other stuff like `Text` instead of having just try-catch blocks everywhere

>long? number = fieldVal.AsPhoneNumber(/* bool validate = */ false);.

Not sure I follow. Could you elaborate more?

**Comment by maririos** [comment]
@tg-msft , @annelo-msft , @kinelski 
 have updated the branch with the approach suggested by Ted of having `AsXXOrNull` for all methods excluding the ones that return string as they already return null if the type was `string`.

I find it kind of odd for `list` and `dictionary` ... but when writing the sample code, I did like the consistency of not having to add try/catch around those two and have all the field extraction logic be the same.

I have removed tests for now until the design is considered better or approved.

**Comment by tg-msft** [comment]
> > long? number = fieldVal.AsPhoneNumber(/* bool validate = */ false);.
> 
> Not sure I follow. Could you elaborate more?

Sorry, that was unintelligible.  That should have been `AsInt64(validate: false)`.  Imagine if we had APIs like:

```C#
public long AsInt64(); // Exists today and throws
public long? AsInt64(bool validate); // Only throws if validate == true so it also makes value return types nullable
public IReadOnlyList<FormField> AsList();
public IReadOnlyList<FormField> AsList(bool validate);
// ...
```

Then customers who didn't want to wrap things in try/catch could write code like:

```C#
long? amount = fieldVal.AsInt64(validate: false);
if (amount != null)
{
    // ...
}
```

The `validate` parameter only saves you a character (a `false` parameter value vs. an `OrNull` suffix), but it makes intellisense less cluttered.  It might be a little less discoverable though.

**Comment by maririos** [comment]
@tg-msft I see the value for the simplification but I worry about having an "overload" that has a different return type for some cases and not others.
For now, and unless someone has strong opinions, I am sticking with `AsXXOrNull` as the name makes it more clear for the user what will happen when using that method.

File: sdk/formrecognizer/Azure.AI.FormRecognizer/src/FieldValue.cs [hunk 5]
```diff
@@ -218,6 +218,23 @@ public long AsInt64()
...
+        /// <summary>
+        /// Gets the value of the field as a <see cref="long"/>.
+        /// If the operation fails, null value will be returned.
+        /// </summary>
+        /// <returns>The value of the field converted to a <see cref="long"/>, or null in case of failure.</returns>
+        public long? AsInt64OrNull()
+        {
+            try
+            {
+                return AsInt64();
```
Comment (annelo-msft) [inline sdk/formrecognizer/Azure.AI.FormRecognizer/src/FieldValue.cs:230]: Is this intended to be a performance optimization to prevent the exception machinery from getting invoked?  If so, it might make sense to do the check inside AsInt64() directly rather than relying on it to throw an exception to return null.

File: sdk/formrecognizer/Azure.AI.FormRecognizer/tests/samples/Sample13_RecognizeInvoicesFromFile.cs [hunk 6]
```diff
@@ -42,103 +42,156 @@ public async Task RecognizeInvoicesFromFile()
 
             if (invoice.Fields.TryGetValue("VendorName", out FormField vendorNameField))
             {
-                if (vendorNameField.Value.ValueType == FieldValueType.String)
+                string vendorName = vendorNameField.Value.AsString();
```
Comment (annelo-msft) [inline sdk/formrecognizer/Azure.AI.FormRecognizer/tests/samples/Sample13_RecognizeInvoicesFromFile.cs:45]: Why would this not be AsStringOrNull() like the others?  Is that because null is a valid option for String?

**Comment by annelo-msft** [comment]
> The main concern users had was that they thought checking the type of the FieldValue was enough for trying our AsType methods. When they got exceptions thrown, there was confusion on the why.

Do we know what their concern is about the exception?  Is it performance?  Is it that they have to catch the exception?

This works, but I have a concern about the semantics -- you can get null either because the field is a different type (i.e. it's a Float, not an Int64) or because the value of the field that is a Float has value of null.  Is null semantically meaningful?  Does it mean the field didn't have a value in the form, the confidence of the ML was too low to populate it, or that the field itself was absent?  It seems like there could feasibly be scenarios for each -- i.e. was my form filled out?  Is this the wrong type of form?  

It's also worth noting that in principle, ppl shouldn't have to check for the type in production ... it's fixed once the model is trained, so that exception would probably only be thrown in development.

Some alternatives, depending on the meaning of a null ...

- Add HasValue() to FormField.
If they're fine checking the FieldValueType, and want to avoid the exception, would they be fine to check FieldValueType AND HasValue()?  As mentioned above, only the latter should need to be checked in production.

- In deserialization, don't add the field to Fields if it has a null value.  
This would depend on the semantics of null.  But if it made sense wrt the semantics, this:

```C#
            if (invoice.Fields.TryGetValue("Items", out FormField itemsField))
            {
                IReadOnlyList<FormField> itemFieldsList = itemsField.Value.AsList();
                if (itemFieldsList != null)
                {
                    foreach (FormField itemField in itemFieldsList)
                    {
                        Console.WriteLine("Item:");

                        IReadOnlyDictionary<string, FormField> itemFields = itemField.Value.AsDictionaryOrNull();
                        if (itemFields != null)
                        {
                            if (itemFields.TryGetValue("Description", out FormField itemDescriptionField))
                            {
                                string itemDescription = itemDescriptionField.Value.AsString();
                                if (!string.IsNullOrEmpty(itemDescription))
                                {
                                    Console.WriteLine($"Description: '{itemDescription}', with confidence {itemDescriptionField.Confidence}");
                                }
                                else
                                {
                                    Console.WriteLine($"Not able to parse Description. Consider using the 'ValueData.Text' property which current value is '{itemDescriptionField.ValueData.Text}'");
                                }
                            }

                  // ...
            }
```

Could become this:

```C#
            if (invoice.Fields.TryGetValue("Items", out FormField itemsField))
            {
                foreach (FormField itemField in itemsField.Value.AsList())
                {
                    IReadOnlyDictionary<string, FormField> itemFields = itemField.Value.AsDictionary();

                    if (itemFields.TryGetValue("Description", out FormField itemDescriptionField))
                    {
                        string itemDescription = itemDescriptionField.Value.AsString();

                        Console.WriteLine($"  Description: '{itemDescription}', with confidence {itemDescriptionField.Confidence}");

                    }
                 // ...
            }
```

Or, if you want to be super safe at runtime in production:

```C#
            if (invoice.Fields.TryGetValue("Items", out FormField itemsField) && itemsField.Value.ValueType == FieldValueType.List))
            {
                // ...
```

File: sdk/formrecognizer/Azure.AI.FormRecognizer/src/FieldValue.cs [hunk 7]
```diff
@@ -218,6 +218,23 @@ public long AsInt64()
...
+            }
+            catch
```
Comment (kinelski) [inline sdk/formrecognizer/Azure.AI.FormRecognizer/src/FieldValue.cs:232]: Maybe we should be more specific about the type of exception we're expecting here (I think only `InvalidOperationException` is expected to happen, right?). Otherwise, if we have an unexpected exception (`NullReferenceException` or `OutOfMemoryException`, for example), we could be potentially hiding a client bug or an error in user code.

File: sdk/formrecognizer/Azure.AI.FormRecognizer/src/FieldValue.cs [hunk 8]
```diff
@@ -346,6 +414,23 @@ public IReadOnlyList<FormField> AsList()
             return fieldList;
         }
 
+        /// <summary>
+        /// Gets the value of the field as a <see cref="IReadOnlyList{T}"/>.
```
Comment (kinelski) [inline sdk/formrecognizer/Azure.AI.FormRecognizer/src/FieldValue.cs:418]: nit: "as a" => "as an".

File: sdk/formrecognizer/Azure.AI.FormRecognizer/src/FieldValue.cs [hunk 9]
```diff
@@ -346,6 +414,23 @@ public IReadOnlyList<FormField> AsList()
...
+        /// If the operation fails, null value will be returned.
+        /// </summary>
+        /// <returns>The value of the field converted to a <see cref="IReadOnlyList{T}"/>, or null in case of failure.</returns>
```
Comment (kinelski) [inline sdk/formrecognizer/Azure.AI.FormRecognizer/src/FieldValue.cs:421]: nit: "to a" => "to an"

File: sdk/formrecognizer/Azure.AI.FormRecognizer/src/FieldValue.cs [hunk 10]
```diff
@@ -373,6 +458,23 @@ public IReadOnlyList<FormField> AsList()
             return fieldDictionary;
         }
 
+        /// <summary>
+        /// Gets the value of the field as a <see cref="Dictionary{TKey, TValue}"/>.
+        /// If the operation fails, null value will be returned.
+        /// </summary>
+        /// <returns>The value of the field converted to a <see cref="Dictionary{TKey, TValue}"/>, or null in case of failure.</returns>
```
Comment (kinelski) [inline sdk/formrecognizer/Azure.AI.FormRecognizer/src/FieldValue.cs:465]: "a Dictionary{TKey, TValue}" => "an IReadOnlyDictionary{TKey, TValue}"

File: sdk/formrecognizer/Azure.AI.FormRecognizer/tests/samples/Sample13_RecognizeInvoicesFromFile.cs [hunk 11]
```diff
@@ -42,103 +42,156 @@ public async Task RecognizeInvoicesFromFile()
...
+                else
+                {
+                    Console.WriteLine($"Not able to parse Customer Name. Consider using the 'ValueData.Text' property which current value is '{customerNameField.ValueData.Text}'");
+                }
             }
 
             if (invoice.Fields.TryGetValue("Items", out FormField itemsField))
             {
-                if (itemsField.Value.ValueType == FieldValueType.List)
+                IReadOnlyList<FormField> itemFieldsList = itemsField.Value.AsList();
```
Comment (kinelski) [inline sdk/formrecognizer/Azure.AI.FormRecognizer/tests/samples/Sample13_RecognizeInvoicesFromFile.cs:71]: Should this be `AsListOrNull`?

**Comment by tg-msft** [comment]
> It's also worth noting that in principle, ppl shouldn't have to check for the type in production ... it's fixed once the model is trained, so that exception would probably only be thrown in development.

Is that what we recommend customers do?  How does that hold up to versioning of the service/model/etc.?  I thought there's a chance that new types of fields get added over time and things returned as string today might come back as FooBars in the future?

File: sdk/formrecognizer/Azure.AI.FormRecognizer/tests/samples/Sample13_RecognizeInvoicesFromFile.cs (hunk 6 above)
Comment (maririos) [reply sdk/formrecognizer/Azure.AI.FormRecognizer/tests/samples/Sample13_RecognizeInvoicesFromFile.cs]: the service doesn't return a `ValueType` when the normalization to a type fails. After I confirmed with the service, this will only happen for types `number, integer, date, time, selectionMark, country, phoneNumber, and gender`.
I added `xxOrNull` for those types.

**Comment by maririos** [comment]
>Is that what we recommend customers do? How does that hold up to versioning of the service/model/etc.? I thought there's a chance that new types of fields get added over time and things returned as string today might come back as FooBars in the future?

I agree. I prefer to always leave the check. For context, the service hasn't done this ^ but that doesn't mean they won't in the future.

**Comment by maririos** [comment]
**Update on the PR**

First, thank you all for brainstorming and reviewing :) I really appreciate it.
Why am I doing this?
- Major pain point: How do we communicate to the customer that the service can recognize a field, but fails to normalize it to a specific type?  currently, we just throw an exception saying `field value is null` and it is not clear to the user what to do. This is backed by the amount of issues/emails that get sent our way with people thinking the SDK failed to parse a field into a type.
    The way other languages manage this is by setting `FormField.Value = null` . Ideally, we should do the same, but this will introduce a breaking change for anyone using `3.0 GA` so the next best thing is to be very clear about what that exception really means.
- Customers complain that in order to get information from a document, their code needs to be full of `try/catch` blocks and makes it very hard to follow.

After all your ideas and me going back to customer feedback, playing with samples, and brainstorming with the other languages, the main candidates to solve the problem were:

  | Pros | Cons
-- | -- | --
HasValue | Follows our Operation pattern where the user can check if the property has a value,   before accesing it. | - Covers pain point 1, but not 2.  - Not easily discoverable. Might cause that in the developer environment user doesn't face a situation where we throw and therefore never learn about the `HasValue` property causing future problems in production.
Add   `AsXxOrNull` and hide existing `AsXx` | - Covers both pain points.     - Easily discoverable as there is no other option.     - Makes it clear to the user that they need to pay attention and manage scenarios where `null` is returned.     - Allows us to fix a bug that was introduced in 3.0 in the AsFloat method (https://github.com/Azure/azure-sdk-for-net/issues/20014) | - Not all methods should use this pattern, only the types that need normalization on the service side and this might be confusing for the user.     - Maintain duplicated code in the SDK.

Because of this, I decided to go with `AsXxOrNull` approach.
@tg-msft @annelo-msft @kinelski  what do you think?

Note about PR. Once design has green light:
- Still needs tests
- Need to clean methods that are not GAd and only expose the `AsXxOrNull` and update respective samples and tests (i.e CountryCode, Gender, SelectionMarkState)
- Need to update all other samples (this will happen later or the PR will explode).

**Comment by maririos** [comment]
FYI @KrzysztofCwalina . Per your comment in the API View tool:

Could you elaborate on why the current approach is not appropiate?

As for the other options:

>we should explore: 1) add property IsNull,

This will be similar to what @annelo-msft proposed of `HasValue`, right? See my comment above with the reasoning why I decided not to go with it. If my priotizaiton was off, please let me know.

> Call the new method ToXxx

For consistency, we will have to apply this change to all methods. Only the methods that have normalized types will return nullable types, so there is no need to add `OrNull` to the method name. 
I am ok with this approach

>Add As<T> method and hide the existing methods.

My first approach to this problem was similar, where I did a generic `TryParse`, but like @tg-msft pointed out, `TryParse implies things that this API isn't doing. If I have a string field with the value "10", I don't think TryParse(out int x) will actually parse the value and give me back an int.`
I can see how `As<T>` might have similar problems.

**Comment by KrzysztofCwalina** [comment]
> My first approach to this problem was similar, where I did a generic TryParse, but like @tg-msft pointed out, TryParse implies things that this API isn't doing. If I have a string field with the value "10", I don't think TryParse(out int x) will actually parse the value and give me back an int.
I can see how As<T> might have similar problems.

I might not be understanding Ted's concern, but it seems to me he was saying that if the value is a string type, not an int type, we will not actually parse the string to an int, which would be surprising given the word "parse" (implying parsing) in the name of the method. As<T> would not have this problem.

**Comment by maririos** [comment]
>I might not be understanding @tg-msft's concern, but it seems to me he was saying that if the value is a string type, not an int type, we will not actually parse the string to an int, which would be surprising given the word "parse" (implying parsing) in the name of the method. As would not have this problem.

Hmm interesting. Let me play with the idea and report back the findings. Thank you!!
